resource/graph_node.go

```diff
@@ -209,6 +239,7 @@ func (w *GraphNode) setNeedsReconfigure(newConfig Config, mustReconfigure bool,
 	}
 	if mustReconfigure {
 		w.needsDependencyResolution = true
+		w.timesReconfigured.Store(0)
```

Comment:
Ah I missed this before; so this means that `SetNeedsUpdate` will _not_ reset the count (I see you're asserting that that is the case in the test, too).

If that's a decision we're making we should document the reasoning somewhere in a comment. With the current design, if a resource failed to reconfigure within the resource configuration timeout after 1 minute, and then its _parent_ reconfigured, the child resource would not get a chance to try reconfiguring again. I can't really imagine a situation where the child resource would suddenly _start_ being able to (re)configure within the resource configuration timeout, so that seems fair to me. 

If you both agree, then I'd just ask for a comment above this line clarifying that decision.

Reply:
Yeaaah, I just figured since I thought `SetNeedsUpdate` is invoked without any conditions changing for the resource to (re)configure properly, it wouldn't make sense to reset the count. I wasn't aware about the parent/child relationship though. But according to your comment it shouldn't be a problem? Idk we can change it or keep it since my knowledge of the system is relatively myopic.

Reply:
>  But according to your comment it shouldn't be a problem?

Yeah, I think what you currently have makes sense but is a decision. @cheukt what are your thoughts?

Reply:
I'd argue that the parent becoming available is a condition that changed, though I may also be misreading something.
q: right now, if a parent failed to re/configure for a few times (so the child also reached the max reconfigure limit) and the user updates the parent config so the parent resource finally starts up properly, will the child reconfigure or be stuck at the max reconfigure limit?

Reply:
Let me think about changing it to (re)config when SetNeedsUpdate is called a little harder + let you guys know if changing it has any issues

Cheuk's q:
```// markResourceForUpdate marks the given resource in the graph to be updated. If it does not exist, a new node
// is inserted. If it does exist, it's properly marked. Once this is done, all information needed to build/reconfigure
// will be available when we call completeConfig.
func (manager *resourceManager) markResourceForUpdate(name resource.Name, conf resource.Config, deps []string) error {
	gNode, hasNode := manager.resources.Node(name)
	if hasNode {
		gNode.SetNewConfig(conf, deps)
		// reset parentage
		for _, parent := range manager.resources.GetAllParentsOf(name) {
			manager.resources.RemoveChild(name, parent)
		}
		return nil
	}
	gNode = resource.NewUnconfiguredGraphNode(conf, deps)
	if err := manager.resources.AddNode(name, gNode); err != nil {
		return errors.Errorf("failed to add new node for unconfigured resource %q: %v", name, err)
	}
	return nil
}```

Might be reading/interpreting this code wrong, but from my understanding here is that the children are removed when the parent is modified and re-added as a new node (thus resetting the attempt count)

Reply:
Ok updated answer: I think only passed the `diff`ed config  is passed to `r.manager.updateResources(ctx, diff)`. So the child would indeed be stuck.

Reply:
if you run a test manually for that scenario, what happens? it'd be good to verify what you think will happen with what actually ends up happening

Reply:
Changing `SetNeedsUpdate` to reset the count fixes the above

Reply:
Cool I'm fine with resetting the count with `SetNeedsUpdate`. Thanks for looking into it.